### PR TITLE
Feature/7 data storage

### LINKED ISSUE #7: Implement Redunda's data storage
SOBotics/Redunda#26 introduces API routes to store files that are shared between instances.

The Java library should be able to track files and sync them automatically. (replace the local file, if the remote is newer)

### PR DESCRIPTION
**Pull request by FelixSFD**
Note: I'll copy this text to the wiki, once this PR is merged.

Closes #7 

---

# How to sync files

### Get the DataService

You can easily get the `DataService` with the same API-key as `PingService`:

```
DataService redundaData = redunda.buildDataService();
```

### Track files for syncing

Each file that should be automatically synchronized needs to be tracked:

```
redundaData.trackFile("./myFile");
```

**Note:** You should __not__ use absolute paths and a path should not contain the String `_slash_`.

### Start synchronizing

The automatic sync can be started like this:

```
redundaData.syncAndStart();
```

This will call `syncFiles()` __synchronously__ (for proper initialization of your bot) and then start the background thread. Alternatively, you can use `start()` if you don't need the files on launch.

If you don't want to sync automatically in background, you can just call `syncFiles()`.

## Upload and download files manually

If you don't want to sync automatically, you can call the methods for uploading and downloading by yourself. They are the same as `syncFiles()` uses. Please have a look at the documentation for more information.

### CODE REVIEW

**Comment by FelixSFD** [comment]
I'll wait until SOBotics/Redunda#35 is merged.

**Comment by FelixSFD** [comment]
Done. It can be reviewed and merged now @ArtOfCode- 

Snapshot build is already deployed. I'll deply a release build once the PR is merged.

File: src/main/java/org/sobotics/redunda/DataService.java [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+		}
+		
+		
+		String url = "https://redunda.sobotics.org/bots/data/"+encodedFilename+"?key="+this.apiKey;
+		URL obj = new URL(url);
+		HttpsURLConnection con = (HttpsURLConnection) obj.openConnection();
+
+		//add request header
+		con.setRequestMethod("POST");
+		con.setRequestProperty("User-Agent", "Redunda Library");
```
Comment (ArtOfCode-) [inline src/main/java/org/sobotics/redunda/DataService.java]: This doesn't conform to the User-Agent spec: `requester-name/requester-version [extra-info]`. Maybe something like `redunda-lib-java/1.7`?

File: src/main/java/org/sobotics/redunda/DataService.java [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+	 * @return An array of objects with information about the stored files
+	 * */
+	public JsonArray getRemoteFileList() throws Throwable {
+		String url = "https://redunda.sobotics.org/bots/data.json?key="+this.apiKey;
+		URL obj = new URL(url);
+		HttpsURLConnection con = (HttpsURLConnection) obj.openConnection();
+
+		//add request header
+		con.setRequestMethod("GET");
+		con.setRequestProperty("User-Agent", "Redunda Library");
```
Comment (ArtOfCode-) [inline src/main/java/org/sobotics/redunda/DataService.java]: Likewise with user agents here

File: src/main/java/org/sobotics/redunda/DataService.java [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+			return null;
+		}
+		
+		String url = "https://redunda.sobotics.org/bots/data/"+encodedFilename+"?key="+this.apiKey;
+		URL obj = new URL(url);
+		HttpsURLConnection con = (HttpsURLConnection) obj.openConnection();
+
+		//add request header
+		con.setRequestMethod("GET");
+		con.setRequestProperty("User-Agent", "Redunda Library");
```
Comment (ArtOfCode-) [inline src/main/java/org/sobotics/redunda/DataService.java]: you know the drill

File: src/main/java/org/sobotics/redunda/DataService.java (hunk 1 above)
Comment (FelixSFD) [reply src/main/java/org/sobotics/redunda/DataService.java]: okay. I'll change that in `PingService` as well

**Comment by FelixSFD** [comment]
@ArtOfCode- done. The user agent will contain the library's version number

**Pull request merged**